src/PostgREST/App.hs

```diff
@@ -188,21 +187,30 @@ app dbStructure proc conf apiRequest =
                     (iPreferRepresentation apiRequest) []
               row <- H.statement (toS $ pjRaw pJson) stm
               let (_, queryTotal, _, body) = extractQueryResult row
-              if contentType == CTSingularJSON
-                 && queryTotal /= 1
-                 && iPreferRepresentation apiRequest == Full
-                then do
-                  HT.condemn
-                  return $ singularityError (toInteger queryTotal)
-              else do
-                let r = contentRangeH 0 (toInteger $ queryTotal-1)
-                          (toInteger <$> if shouldCount then Just queryTotal else Nothing)
-                    s = if iPreferRepresentation apiRequest == Full
-                          then status200
-                          else status204
-                return $ if iPreferRepresentation apiRequest == Full
-                  then responseLBS s [toHeader contentType, r] (toS body)
-                  else responseLBS s [r] ""
+
+                  updateIsNoOp       = S.null $ pjKeys pJson
+                  contentRangeHeader = contentRangeH shouldCount 0 (queryTotal - 1) queryTotal
+                  minimalHeaders     = [contentRangeHeader]
+                  fullHeaders        = toHeader contentType : minimalHeaders
+
+                  status | queryTotal == 0 && not updateIsNoOp      = status404
+                         | iPreferRepresentation apiRequest == Full = status200
+                         | otherwise                                = status204
+
+              case (contentType, queryTotal, iPreferRepresentation apiRequest) of
```

Comment:
queryTotal is only used by 1 match, so how about expressing this as:
```haskell
case (contentType, iPreferRepresentation apiRequest) of
  (CTSingularJSON, Full)
    | queryTotal == 1 -> return $ responseLBS status fullHeaders (toS body)
    | otherwise -> HT.condemn >> return (singularityError queryTotal)

  (_, Full) ->
    return $ responseLBS status fullHeaders (toS body)

  (_, _) ->
    return $ responseLBS status minimalHeaders mempty
```

Reply:
Changed to suggested.